pkg/scheduler/asks_finder.go

```diff
@@ -64,15 +64,12 @@ func (m *Scheduler) findAllocationAsks(partitionTotalResource *resources.Resourc
 // sorting type for the parent queue.
 // Stopped queues will be filtered out at a later stage.
 func sortSubqueuesFromQueue(parentQueue *SchedulingQueue) []*SchedulingQueue {
-    parentQueue.lock.RLock()
-    defer parentQueue.lock.RUnlock()
```

Comment:
Doubt about removing this lock.
It might be possible during the same time when the scheduler doing the queues sorting, another allocation added/removed to a leaf queue, which invalidates the sorting result, isn't it?

Reply:
Sorting is just one step and the sort can be invalidated already. 
- Getting a copy of the child list is still locked
- After we sort the queue list is passed back. At that point in time we already have released the lock and allocations will occur.
- the child queues which are sorted were never locked.

Sorting also happens on two field that are not lock protected: ProposingResource for the scheduling side and the cached GuaranteedResource field. Both are accessed outside locks. So locking of the parent queue could only have had an indirect effect preventing modifications on the ProposingResource but never on the cached GuaranteedResource.

Based on all that I removed the lock

Reply:
I was quite surprised that the accessing of queue resources were not lock protected, I just went over the related code again. Now I see why.  We are not using real multi-thread allocating. In tryBatchAllocation, we can possibly launch a few threads, but we wait on them to return proposals. So sort queue and modify queue resources are not happening concurrently. I think we are safe to remove this lock now.

But we may need to revise this model sometime later. Usually, we run allocation threads concurrently and produce proposals and the main scheduler thread just consumes them by either committing or rejecting them.

Reply:
I agree when we start moving towards multithreaded allocations we need to go over the whole locking with a fine-tooth comb. We might get a different set of locks when we get to that.